source/server/connection_handler_impl.h

```diff
@@ -336,6 +347,16 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
   std::list<std::pair<Network::Address::InstanceConstSharedPtr, ActiveListenerDetails>> listeners_;
   std::atomic<uint64_t> num_handler_connections_{};
   bool disable_listeners_;
+
+  using SocketMetadataPair =
+      std::pair<Network::ConnectionSocketPtr, const envoy::config::core::v3::Metadata>;
+
+  // The pending sockets waiting for filter chains to be rebuilt. Multiple listeners might request
+  // rebuilding the same filter chain.
+  absl::flat_hash_map<envoy::config::listener::v3::FilterChain,
+                      absl::flat_hash_map<std::string, std::list<SocketMetadataPair>>, MessageUtil,
+                      MessageUtil>
+      pending_sockets_;
```

Comment:
Conceptually this map is working.
I am wondering if this huge map have to put in `ConnectionHandler`.
You always has the `listener_name` whenever you have `FilterChain` message.

Why don't you 
1. set the top level key as listener_name
2. maintain the 2nd level map<FilterChain, std::list<SocketMetadataPair>> into per listener?

The gain is to delegate the work to underlying active listener

Reply:
My reason that I put filter chain as the first level key of ```pending_sockets``` is that we know the set of listeners that are requesting the same filter chain for rebuilding. 
If we change the listener as the top level key, we have to loop through the all listeners in the pending_sockets map to check if that listener requested one particular filter chain.

Reply:
Fundamentally, Listener A's filter chain FC1 is not the same as Listener B's filter chain FC1.
To identify the impacted connection, or impacted filter chain, you do need both FC and listener name anytime. 

I think for look up, we actually need key as pair<ListenerName, filterchain> *if not b/c garbage collection looping though all the FC/Listener"
Does above GC exist?

Reply:
> Fundamentally, Listener A's filter chain FC1 is not the same as Listener B's filter chain FC1.
Given this information, I think that ```pending_sockets``` using ```pair<filter_chain, listener>``` as key is fine.

Reply:
Thanks! Thanks a great catch, which will save the iterate. Will Update!

Reply:
> I am wondering if this huge map have to put in ConnectionHandler

You are right, I think its better to store the pending_sockets directly inside each active_tcp_listener